snapcraft/commands/extensions.py

```diff
@@ -0,0 +1,95 @@
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Snapcraft lifecycle commands."""
+
+import abc
+import textwrap
+from typing import Dict, List
+
+import tabulate
+from craft_cli import BaseCommand, emit
+from overrides import overrides
+from pydantic import BaseModel
+
+from snapcraft import extensions
+from snapcraft_legacy.internal.project_loader import (
+    find_extension,
+    supported_extension_names,
+)
+
+
+class ExtensionModel(BaseModel):
+    """Extension model for presentation."""
+
+    name: str
+    bases: List[str]
+
+    def marshal(self) -> Dict[str, str]:
+        """Marshal model into a dictionary for presentation."""
+        return {
+            "Extension name": self.name,
+            "Supported bases": ", ".join(sorted(self.bases)),
+        }
+
+
+class ListExtensionsCommand(BaseCommand, abc.ABC):
+    """A command to list the available extensions."""
+
+    name = "list-extensions"
+    help_msg = "List available extensions"
+    overview = textwrap.dedent(
+        """
+        List the available extensions and the bases it can work on.
+        """
+    )
+
+    @overrides
+    def run(self, parsed_args):
+        extension_presentation: Dict[str, ExtensionModel] = {}
+
+        # New extensions.
+        for extension_name in extensions.registry.get_extension_names():
+            extension_class = extensions.registry.get_extension_class(extension_name)
+            extension_bases = list(extension_class.get_supported_bases())
+            extension_presentation[extension_name] = ExtensionModel(
+                name=extension_name, bases=extension_bases
+            )
+
+        # Extensions from snapcraft_legacy.
+        for extension_name in supported_extension_names():
+            extension_class = find_extension(extension_name)
+            extension_name = extension_name.replace("_", "-")
+            extension_bases = list(extension_class.get_supported_bases())
+            if extension_name in extension_presentation:
+                extension_presentation[extension_name].bases += extension_bases
+            else:
+                extension_presentation[extension_name] = ExtensionModel(
+                    name=extension_name, bases=extension_bases
+                )
+
+        printable_extensions = sorted(
+            [v.marshal() for v in extension_presentation.values()],
+            key=lambda d: d["Extension name"],
+        )
+        emit.message(tabulate.tabulate(printable_extensions, headers="keys"))
+
+
+class ExtensionsCommand(ListExtensionsCommand, abc.ABC):
```

Comment:
craft-cli should have first-class support for command aliases.

Reply:
I created canonical/craft-cli#74 for that